server/game/GameActions/RevealCards.js

```diff
@@ -0,0 +1,107 @@
+const GameAction = require('./GameAction');
+const HandlerGameActionWrapper = require('./HandlerGameActionWrapper');
+const AcknowledgeRevealCardsPrompt = require('../gamesteps/AcknowledgeRevealCardsPrompt');
+
+class RevealCards extends GameAction {
+    constructor() {
+        super('revealCards');
+        this.defaultWhileRevealed = new HandlerGameActionWrapper({ handler: context => {
+                if(context.revealed.length > 0) {
+                    context.game.queueStep(new AcknowledgeRevealCardsPrompt(context.game, context.revealed, context.player))
+                }
+            }
+        });
+    }
+
+    canChangeGameState({ cards }) {
+        cards = Array.isArray(cards) ? cards : [cards];
+        return cards.length > 0 && cards.some(card => this.isInHiddenArea(card));
+    }
+
+    createEvent({ cards, player, whileRevealed, context }) {
+        const allPlayers = context.game.getPlayers();
+        const eventParams = {
+            player,
+            opponents: allPlayers.filter(p => p !== player),
+            cards: Array.isArray(cards) ? cards : [cards],
+            source: context.source
+        }
+        return this.event('onCardsRevealed', eventParams, event => {
+            const whileRevealedGameAction = whileRevealed || this.defaultWhileRevealed;
+            const playerRevealFunc = (card, player) => event.cards.includes(card) && player === event.player;
+            const allPlayersRevealFunc = card => event.cards.includes(card);
+
+            // Initially make cards visible to the revealing player before the actual "reveal" event. This ensures any interrupts to 'onCardRevealed' will actually show the card (eg. Alla Tyrell)
+            context.game.cardVisibility.addRule(playerRevealFunc);
+            
+            for(let card of event.cards) {
+                const revealEventParams = {
+                    card,
+                    cardStateWhenRevealed: card.createSnapshot(),
+                    player: event.player,
+                    opponents: event.opponents,
+                    source: event.source
+                }
+                
+                event.thenAttachEvent(this.event('onCardRevealed', revealEventParams, () => {}));
+            }
+            context = Object.assign({ revealed: event.cards, player: event.player, opponents: event.opponents }, context);
```

Comment:
I'm not sure we want to do this. This is creating a plain Javascript object, whereas the `context` being passed in is likely an `AbilityContext` object.

While it might not matter in most cases, I think it could be a problem if we have a `whileRevealed` game action, which could be using methods that would not be copied over this way.

I think I'd rather see us create a new context object (as we do in the ThenClauseAbility), or do something simple like:
```
context.revealData = { ... }
```

Or even just:
```
context.revealedCards = event.cards
```

since `player` and `opponents` are not used in here and may not be needed in the `whileRevealed` action.

Reply:
This has been fixed/addressed in e2661e091c5a94eb156a653cf08385a597cf4e74. I'd recommend resolving this comment.